### PR TITLE
Added custom crops

### PR DESCRIPTION
**Pull request by anirudh-challa**
This PR adds `custom_crop` a boolean value to the constructor, if True it reads the `crop` field from image manifest dataframe. It also adds `crop_col` a variable to define which column to read from and by default it is the `crop` column.

### CODE REVIEW

File: keras_trainer/data_generators.py [hunk 1]
```diff
@@ -41,6 +42,7 @@ def set_processing_attrs(self,
                 to use for random transformations and normalization.
             random_crop_size: Size of the random crop. Either a percentage of the original image (0,1) that will do square
                 crop, a fixed size (tuple), or integer where the value will set equally to both dimensions.
+            custom_crop: If True, will crop images according to the manifest file.
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]: Correct me if I'm wrong, but we haven't mentioned the term "manifest files" in `keras-trainer` yet, so: 

```suggestion
            custom_crop: If True, will crop images according to the dataframe's crop coordinates information contained in `z_col`.
```

File: keras_trainer/data_generators.py [hunk 2]
```diff
@@ -138,6 +141,23 @@ def random_crop_parameters(img, crop_size, crop_mode='size'):
 
         return x, y, x + tw, y + th
 
+    @staticmethod
+    def apply_custom_crop(image, crop_coordinates):
+        """
+        The format of the image cropping coordinates should be [x, y, width, height] and PIL cropping format is
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]: Maybe to better reference our expected arguments:
```suggestion
        The format of the image `crop_coordinates` should be [x, y, width, height] and PIL cropping format is
```

File: keras_trainer/data_generators.py [hunk 3]
```diff
@@ -26,6 +26,7 @@ class EnhancedBatchFromFilesMixin(BatchFromFilesMixin):
     def set_processing_attrs(self,
                              image_data_generator,
                              random_crop_size,
+                             custom_crop,
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]: The same way you're referencing `custom_crop` before `random_crop_size` in `_get_batches_of_transformed_samples()`, by not defining it here before too?
```suggestion
                             custom_crop,
                             random_crop_size,
```

File: keras_trainer/data_generators.py [hunk 4]
```diff
@@ -431,6 +456,7 @@ class EnhancedDataFrameIterator(EnhancedBatchFromFilesMixin, EnhancedIterator):
         x_col: string, column in `dataframe` that contains the filenames (or
             absolute paths if `directory` is `None`).
         y_col: string or list, column/s in `dataframe` that has the target data.
+        z_col: list or  none, column in `dataframe` that has the custom crop coordinates.
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]: What do you mean by "none"? The column in `dataframe` can be empty, but `z_col` will always be a list (of values or empty), right?

```suggestion
        z_col: list, (optional) column in `dataframe` that contains the custom crop coordinates.
```

File: keras_trainer/data_generators.py [hunk 5]
```diff
@@ -302,6 +325,7 @@ def __init__(self,
                  image_data_generator,
                  iterator_mode=None,
                  random_crop_size=None,
+                 custom_crops=False,
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]: If this variable contains the crops, should it be initialized to `None` instead? 
```suggestion
                 custom_crops=None,
                 random_crop_size=None,
```

File: keras_trainer/data_generators.py [hunk 6]
```diff
@@ -320,6 +344,7 @@ def __init__(self,
                  dtype='float32'):
         super(EnhancedDirectoryIterator, self).set_processing_attrs(image_data_generator,
                                                                     random_crop_size,
+                                                                    custom_crops,
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]:
```suggestion
                                                                    custom_crops,
                                                                    random_crop_size,
```

File: keras_trainer/data_generators.py [hunk 7]
```diff
@@ -550,6 +579,7 @@ def __init__(self,
         if class_mode not in ["input", "multi_output", "raw", "probabilistic", None]:
             self.classes = self.get_classes(df, y_col)
         self.filenames = df[x_col].tolist()
+        self.crops = df[z_col].tolist()
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]: Should you check first if `z_col` exists?

File: keras_trainer/data_generators.py [hunk 8]
```diff
@@ -712,12 +742,14 @@ class EnhancedImageDataGenerator(ImageDataGenerator):
...
 
     This is a modification of the Keras class ImageDataGenerator that includes some extra functionalities.
-    The image data generator has a new option `random_crop_size` that if specified will perform random crops prior to
+    The image data generator has a new options `random_crop_size` and `custom_crop` that if specified will perform random crops prior to
     resize the image to the specified target size.
 
     - random_crop_size:  Size of the random crop. Either a percentage of the original image (0,1) that will do square
         crop, a fixed size (tuple), or integer where the value will set equally to both dimensions.
         target_size: tuple of integers, dimensions to resize input images to.
+    - custom_crop: A boolean, if True custom crops will be done according to the value in 'z_col'. This has been
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]: I believe Albert uses ` :
```suggestion
    - custom_crop: A boolean, if True custom crops will be done according to the value in `z_col`. This has been
```

File: keras_trainer/data_generators.py [hunk 9]
```diff
@@ -712,12 +742,14 @@ class EnhancedImageDataGenerator(ImageDataGenerator):
...
+        implemented only for  `flow_from_dataframe`.
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]:
```suggestion
        implemented only for `flow_from_dataframe()`.
```

File: tests/files/catdog/train_data.json [hunk 10]
```diff
@@ -2,31 +2,37 @@
...
-    "class_name": "dog"
+    "class_name": "dog",
+    "crop": null
   },
   {
     "filename": "dog/dog-3.jpg",
     "class_probabilities": [0.1, 0.9],
-    "class_name": "dog"
+    "class_name": "dog",
+    "crop": null
```
Comment (adriaromero) [inline tests/files/catdog/train_data.json]: The behaviour either if the manifest contains "crop" or not should be the same, right? 

Should we add an image case without "crop" field?

File: tests/test_data_generators.py [hunk 11]
```diff
@@ -89,3 +89,28 @@ def test_enhanced_image_data_generator_random_crop(train_catdog_dataset_path, tr
...
+                                            directory=train_catdog_dataset_path,
+                                            x_col="filename",
+                                            y_col="class_probabilities",
+                                            z_col="crop",
+                                            batch_size=1,
+                                            target_size=(256, 256))
+    x, y = datagen.next()
+    assert x.shape == (1, 256, 256, 3)
+    assert len(y) == 1
+    img = Image.open('tests/files/catdog/train/cat/cat-1.jpg')
```
Comment (adriaromero) [inline tests/test_data_generators.py]: To be imported from the `conftest.py ` file?

File: keras_trainer/data_generators.py (hunk 5 above)
Comment (anirudh-challa) [reply keras_trainer/data_generators.py]: This variable is a boolean if True, crops are done. So I would keep the default as `False`.

**Comment by anirudh-challa** [comment]
Okay, sure. Will call it `crop_column`.

File: keras_trainer/data_generators.py [hunk 12]
```diff
@@ -431,6 +457,7 @@ class EnhancedDataFrameIterator(EnhancedBatchFromFilesMixin, EnhancedIterator):
         x_col: string, column in `dataframe` that contains the filenames (or
             absolute paths if `directory` is `None`).
         y_col: string or list, column/s in `dataframe` that has the target data.
+        crop_column: list, (optional) column in `dataframe` that contains the custom crop coordinates.
```
Comment (adriaromero) [inline keras_trainer/data_generators.py]: Following the `x_col` and `y_col` terminology, should we do 
```suggestion
        crop_col: list, (optional) column in `dataframe` that contains the custom crop coordinates.
```
?

Comment (anirudh-challa) [reply keras_trainer/data_generators.py]: I'm fine either way but if this change has to be made it has to be done in at least 7 or 8 places unless you think it really matters.

**Pull request merged**